hedera-node/src/main/java/com/hedera/services/store/models/Account.java

```diff
@@ -156,56 +154,164 @@ public void decrementUsedAutomaticAssocitions() {
 		setAlreadyUsedAutomaticAssociations(--count);
 	}
 
-	public void associateWith(List<Token> tokens, int maxAllowed, boolean automaticAssociation) {
-		final var alreadyAssociated = associatedTokens.size();
-		final var proposedNewAssociations = tokens.size() + alreadyAssociated;
-		validateTrue(proposedNewAssociations <= maxAllowed, TOKENS_PER_ACCOUNT_LIMIT_EXCEEDED);
+	public EntityNumPair getLastAssociatedToken() {
+		return lastAssociatedToken;
+	}
+
+	public void setLastAssociatedToken(final EntityNumPair lastAssociatedToken) {
+		this.lastAssociatedToken = lastAssociatedToken;
+	}
+
+	public int getNumAssociations() {
+		return numAssociations;
+	}
+
+	public void setNumAssociations(final int numAssociations) {
+		if (numAssociations < 0) {
+			// not possible
+			this.numAssociations = 0;
+		} else {
+			this.numAssociations = numAssociations;
+		}
+	}
+
+	public int getNumZeroBalances() {
+		return numZeroBalances;
+	}
+
+	public void setNumZeroBalances(final int numZeroBalances) {
+		if (numZeroBalances < 0) {
+			// not possible
+			this.numZeroBalances = 0;
+		} else {
+			this.numZeroBalances = numZeroBalances;
+		}
+	}
 
+	public List<TokenRelationship> associateWith(
+			List<Token> tokens,
+			TypedTokenStore tokenStore,
+			boolean isAutomaticAssociation,
+			boolean shouldEnableRelationship) {
 		final Set<Id> uniqueIds = new HashSet<>();
+		List<TokenRelationship> tokenRelationshipsToPersist = new ArrayList<>();
+		var currKey = lastAssociatedToken;
+		TokenRelationship prevRel = currKey.equals(EntityNumPair.MISSING_NUM_PAIR) ?
+				null : tokenStore.getLatestTokenRelationship(this);
 		for (var token : tokens) {
 			final var tokenId = token.getId();
-			validateFalse(associatedTokens.contains(tokenId), TOKEN_ALREADY_ASSOCIATED_TO_ACCOUNT);
-			if (automaticAssociation) {
+			validateTrue(tokenStore.getMerkleTokenRelationship(token, this) == null,
```

Comment:
Given how often this is repeated, might improve readability to have a `tokenStore.hasAssociation(token, this)` method to call.